Nightly crashing on startup

### Describe your problem!

Several users report, that nightly 2023.01.03 is crashing on startup.
I can reproduce on my device.

Seems to be a permission issue for existing installations.

### How to reproduce?

Start c:geo after upgrading to 2023.01.03 nightly.

### Actual result after these steps?

```
2023-01-03 23:13:50.583 13544-13544 Andro...time pid-13544        E  FATAL EXCEPTION: main
                                                                     Process: cgeo.geocaching, PID: 13544
                                                                     java.lang.RuntimeException: Unable to resume activity {cgeo.geocaching/cgeo.geocaching.MainActivity}: java.lang.SecurityException: Permission android.permission.READ_EXTERNAL_STORAGE denied for package cgeo.geocaching
                                                                     	at android.app.ActivityThread.performResumeActivity(ActivityThread.java:5365)
                                                                     	at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:5431)
                                                                     	at android.app.servertransaction.ResumeActivityItem.execute(ResumeActivityItem.java:54)
                                                                     	at android.app.servertransaction.ActivityTransactionItem.execute(ActivityTransactionItem.java:45)
                                                                     	at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:176)
                                                                     	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:97)
                                                                     	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2571)
                                                                     	at android.os.Handler.dispatchMessage(Handler.java:106)
                                                                     	at android.os.Looper.loopOnce(Looper.java:226)
                                                                     	at android.os.Looper.loop(Looper.java:313)
                                                                     	at android.app.ActivityThread.main(ActivityThread.java:8741)
                                                                     	at java.lang.reflect.Method.invoke(Native Method)
                                                                     	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:571)
                                                                     	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1067)
                                                                     Caused by: java.lang.SecurityException: Permission android.permission.READ_EXTERNAL_STORAGE denied for package cgeo.geocaching
                                                                     	at android.os.Parcel.createExceptionOrNull(Parcel.java:3023)
                                                                     	at android.os.Parcel.createException(Parcel.java:3007)
                                                                     	at android.os.Parcel.readException(Parcel.java:2990)
                                                                     	at android.os.Parcel.readException(Parcel.java:2932)
                                                                     	at android.app.IWallpaperManager$Stub$Proxy.getWallpaperWithFeature(IWallpaperManager.java:1996)
                                                                     	at android.app.WallpaperManager$Globals.getCurrentWallpaperLocked(WallpaperManager.java:1073)
                                                                     	at android.app.WallpaperManager$Globals.peekWallpaperBitmap(WallpaperManager.java:936)
                                                                     	at android.app.WallpaperManager$Globals.peekWallpaperBitmap(WallpaperManager.java:863)
                                                                     	at android.app.WallpaperManager.getDrawable(WallpaperManager.java:1380)
                                                                     	at cgeo.geocaching.MainActivity.onResume(MainActivity.java:459)
                                                                     	at android.app.Instrumentation.callActivityOnResume(Instrumentation.java:1531)
                                                                     	at android.app.Activity.performResume(Activity.java:8721)
                                                                     	at android.app.ActivityThread.performResumeActivity(ActivityThread.java:5338)
                                                                     	at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:5431) 
                                                                     	at android.app.servertransaction.ResumeActivityItem.execute(ResumeActivityItem.java:54) 
                                                                     	at android.app.servertransaction.ActivityTransactionItem.execute(ActivityTransactionItem.java:45) 
                                                                     	at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:176) 
                                                                     	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:97) 
                                                                     	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2571) 
                                                                     	at android.os.Handler.dispatchMessage(Handler.java:106) 
                                                                     	at android.os.Looper.loopOnce(Looper.java:226) 
                                                                     	at android.os.Looper.loop(Looper.java:313) 
                                                                     	at android.app.ActivityThread.main(ActivityThread.java:8741) 
                                                                     	at java.lang.reflect.Method.invoke(Native Method) 
                                                                     	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:571) 
                                                                     	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1067) 
                                                                     Caused by: android.os.RemoteException: Remote stack trace:
                                                                     	at android.os.storage.StorageManager.checkPermissionAndAppOp(StorageManager.java:1915)
                                                                     	at android.os.storage.StorageManager.checkPermissionAndAppOp(StorageManager.java:1892)
                                                                     	at android.os.storage.StorageManager.checkPermissionAndAppOp(StorageManager.java:1959)
                                                                     	at android.os.storage.StorageManager.checkExternalStoragePermissionAndAppOp(StorageManager.java:2075)
                                                                     	at android.os.storage.StorageManager.checkPermissionReadImages(StorageManager.java:2041)
2023-01-03 23:13:50.700  1408-1453  Windo...ager pid-1408         E  win=Window{8743936 u0 Splash Screen cgeo.geocaching EXITING} destroySurfaces: appStopped=false cleanupOnResume=false win.mWindowRemovalAllowed=true win.mRemoveOnExit=true win.mViewVisibility=0 caller=com.android.server.wm.WindowState.onExitAnimationDone:5917 com.android.server.wm.ActivityRecord$$ExternalSyntheticLambda11.accept:2 java.util.ArrayList.forEach:1262 com.android.server.wm.ActivityRecord.onAnimationFinished:8440 com.android.server.wm.ActivityRecord.postApplyAnimation:6146 com.android.server.wm.ActivityRecord.commitVisibility:6118 com.android.server.wm.ActivityRecord.commitVisibility:6122 
2023-01-03 23:13:53.343 13704-13704 Andro...time pid-13704        E  FATAL EXCEPTION: main
                                                                     Process: cgeo.geocaching, PID: 13704
                                                                     java.lang.RuntimeException: Unable to resume activity {cgeo.geocaching/cgeo.geocaching.MainActivity}: java.lang.SecurityException: Permission android.permission.READ_EXTERNAL_STORAGE denied for package cgeo.geocaching
                                                                     	at android.app.ActivityThread.performResumeActivity(ActivityThread.java:5365)
                                                                     	at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:5431)
                                                                     	at android.app.servertransaction.ResumeActivityItem.execute(ResumeActivityItem.java:54)
                                                                     	at android.app.servertransaction.ActivityTransactionItem.execute(ActivityTransactionItem.java:45)
                                                                     	at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:176)
                                                                     	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:97)
                                                                     	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2571)
                                                                     	at android.os.Handler.dispatchMessage(Handler.java:106)
                                                                     	at android.os.Looper.loopOnce(Looper.java:226)
                                                                     	at android.os.Looper.loop(Looper.java:313)
                                                                     	at android.app.ActivityThread.main(ActivityThread.java:8741)
                                                                     	at java.lang.reflect.Method.invoke(Native Method)
                                                                     	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:571)
                                                                     	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1067)
                                                                     Caused by: java.lang.SecurityException: Permission android.permission.READ_EXTERNAL_STORAGE denied for package cgeo.geocaching
                                                                     	at android.os.Parcel.createExceptionOrNull(Parcel.java:3023)
                                                                     	at android.os.Parcel.createException(Parcel.java:3007)
                                                                     	at android.os.Parcel.readException(Parcel.java:2990)
                                                                     	at android.os.Parcel.readException(Parcel.java:2932)
                                                                     	at android.app.IWallpaperManager$Stub$Proxy.getWallpaperWithFeature(IWallpaperManager.java:1996)
                                                                     	at android.app.WallpaperManager$Globals.getCurrentWallpaperLocked(WallpaperManager.java:1073)
                                                                     	at android.app.WallpaperManager$Globals.peekWallpaperBitmap(WallpaperManager.java:936)
                                                                     	at android.app.WallpaperManager$Globals.peekWallpaperBitmap(WallpaperManager.java:863)
                                                                     	at android.app.WallpaperManager.getDrawable(WallpaperManager.java:1380)
                                                                     	at cgeo.geocaching.MainActivity.onResume(MainActivity.java:459)
                                                                     	at android.app.Instrumentation.callActivityOnResume(Instrumentation.java:1531)
                                                                     	at android.app.Activity.performResume(Activity.java:8721)
                                                                     	at android.app.ActivityThread.performResumeActivity(ActivityThread.java:5338)
                                                                     	at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:5431) 
                                                                     	at android.app.servertransaction.ResumeActivityItem.execute(ResumeActivityItem.java:54) 
                                                                     	at android.app.servertransaction.ActivityTransactionItem.execute(ActivityTransactionItem.java:45) 
                                                                     	at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:176) 
                                                                     	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:97) 
                                                                     	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2571) 
                                                                     	at android.os.Handler.dispatchMessage(Handler.java:106) 
                                                                     	at android.os.Looper.loopOnce(Looper.java:226) 
                                                                     	at android.os.Looper.loop(Looper.java:313) 
                                                                     	at android.app.ActivityThread.main(ActivityThread.java:8741) 
                                                                     	at java.lang.reflect.Method.invoke(Native Method) 
                                                                     	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:571) 
                                                                     	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1067) 
                                                                     Caused by: android.os.RemoteException: Remote stack trace:
                                                                     	at android.os.storage.StorageManager.checkPermissionAndAppOp(StorageManager.java:1915)
                                                                     	at android.os.storage.StorageManager.checkPermissionAndAppOp(StorageManager.java:1892)
                                                                     	at android.os.storage.StorageManager.checkPermissionAndAppOp(StorageManager.java:1959)
                                                                     	at android.os.storage.StorageManager.checkExternalStoragePermissionAndAppOp(StorageManager.java:2075)
                                                                     	at android.os.storage.StorageManager.checkPermissionReadImages(StorageManager.java:2041)
```

### Expected result after these steps?

_No response_

### Reproducible

Yes

### c:geo Version

2023.01.03-NB

### System information

_No response_

### Additional Information

_No response_

--- a/main/src/main/java/cgeo/geocaching/CacheDetailActivity.java
+++ b/main/src/main/java/cgeo/geocaching/CacheDetailActivity.java
@@ -243,10 +243,8 @@ public class CacheDetailActivity extends TabbedViewPagerActivity
 
     private final EnumSet<TrackableBrand> processedBrands = EnumSet.noneOf(TrackableBrand.class);
 
-    private static final String CONTACT_USER_KEY = "user";
-
-    private final PermissionAction openContactCardAction = PermissionAction.register(this, PermissionContext.SEARCH_USER_IN_CONTACTS, bundle -> {
-        new ContactsHelper(this).openContactCard(bundle.getString(CONTACT_USER_KEY));
+    private final PermissionAction<String> openContactCardAction = PermissionAction.register(this, PermissionContext.SEARCH_USER_IN_CONTACTS, user -> {
+        new ContactsHelper(this).openContactCard(user);
     });
 
     @Override
@@ -599,9 +597,7 @@ public class CacheDetailActivity extends TabbedViewPagerActivity
 
     @Override
     public void showContactCard(final String userName) {
-        final Bundle param = new Bundle();
-        param.putString(CONTACT_USER_KEY, userName);
-        openContactCardAction.launch(param);
+        openContactCardAction.launch(userName);
     }
 
     private abstract static class AbstractWaypointModificationCommand extends AbstractCommand {

--- a/main/src/main/java/cgeo/geocaching/CompassActivity.java
+++ b/main/src/main/java/cgeo/geocaching/CompassActivity.java
@@ -64,7 +64,7 @@ public class CompassActivity extends AbstractActionBarActivity {
     private final TextSpinner<DirectionData.DeviceOrientation> deviceOrientationMode = new TextSpinner<>();
     private CompassActivityBinding binding;
 
-    private final PermissionAction askLocationPermissionAction = PermissionAction.register(this, PermissionContext.LOCATION, b -> {
+    private final PermissionAction<Void> askLocationPermissionAction = PermissionAction.register(this, PermissionContext.LOCATION, b -> {
         binding.locationStatus.updatePermissions();
     });
 

--- a/main/src/main/java/cgeo/geocaching/InstallWizardActivity.java
+++ b/main/src/main/java/cgeo/geocaching/InstallWizardActivity.java
@@ -69,7 +69,7 @@ public class InstallWizardActivity extends AppCompatActivity {
 
     private static final int REQUEST_CODE_WIZARD_GC = 0x7167;
 
-    private final PermissionAction askLocationPermissionAction = PermissionAction.register(this, PermissionContext.LOCATION, b -> gotoNext());
+    private final PermissionAction<Void> askLocationPermissionAction = PermissionAction.register(this, PermissionContext.LOCATION, b -> gotoNext());
 
     // dialog elements
     private ImageView logo = null;

--- a/main/src/main/java/cgeo/geocaching/MainActivity.java
+++ b/main/src/main/java/cgeo/geocaching/MainActivity.java
@@ -58,6 +58,7 @@ import android.content.Context;
 import android.content.Intent;
 import android.content.res.Configuration;
 import android.database.Cursor;
+import android.graphics.drawable.Drawable;
 import android.os.Bundle;
 import android.os.Handler;
 import android.os.Message;
@@ -114,9 +115,8 @@ public class MainActivity extends AbstractBottomNavigationActivity {
 
     private BackupUtils backupUtils = null;
 
-    private final PermissionAction askLocationPermissionAction = PermissionAction.register(this, PermissionContext.LOCATION, b -> {
-        binding.locationStatus.updatePermissions();
-    });
+    private final PermissionAction<Void> askLocationPermissionAction = PermissionAction.register(this, PermissionContext.LOCATION, b -> binding.locationStatus.updatePermissions());
+    private final PermissionAction<Void> askShowWallpaperPermissionAction = PermissionAction.register(this, PermissionContext.SHOW_WALLPAPER, b -> setWallpaper());
 
     private static final class UpdateUserInfoHandler extends WeakReferenceHandler<MainActivity> {
 
@@ -307,6 +307,10 @@ public class MainActivity extends AbstractBottomNavigationActivity {
                 this.askLocationPermissionAction.launch(null);
             });
 
+            if (Settings.isWallpaper()) {
+                askShowWallpaperPermissionAction.launch();
+            }
+
         }
 
         if (Log.isEnabled(Log.LogLevel.DEBUG)) {
@@ -456,7 +460,7 @@ public class MainActivity extends AbstractBottomNavigationActivity {
             updateUserInfoHandler.sendEmptyMessage(-1);
             cLog.add("perm");
 
-            ((ImageView) findViewById(R.id.background)).setImageDrawable(Settings.isWallpaper() ? WallpaperManager.getInstance(this).getDrawable() : null);
+            setWallpaper();
 
             init();
         }
@@ -464,6 +468,12 @@ public class MainActivity extends AbstractBottomNavigationActivity {
         if (Log.isEnabled(Log.LogLevel.DEBUG)) {
             binding.getRoot().post(() -> Log.d("Post after MainActivity.onResume"));
         }
+    }
+
+    private void setWallpaper() {
+        final Drawable wallpaper =  (Settings.isWallpaper() && PermissionContext.SHOW_WALLPAPER.hasAllPermissions()) ?
+                WallpaperManager.getInstance(this).getDrawable() : null;
+            ((ImageView) findViewById(R.id.background)).setImageDrawable(wallpaper);
     }
 
     @Override

--- a/main/src/main/java/cgeo/geocaching/permission/PermissionAction.java
+++ b/main/src/main/java/cgeo/geocaching/permission/PermissionAction.java
@@ -5,20 +5,26 @@ import cgeo.geocaching.ui.TextParam;
 import cgeo.geocaching.ui.dialog.SimpleDialog;
 import cgeo.geocaching.utils.ActivitySavedState;
 import cgeo.geocaching.utils.Log;
+import cgeo.geocaching.utils.ParcelableValue;
 
-import android.app.Activity;
+import android.content.Context;
 import android.content.Intent;
 import android.net.Uri;
-import android.os.Bundle;
 import android.provider.Settings;
 
 import androidx.activity.ComponentActivity;
 import androidx.activity.result.ActivityResultLauncher;
 import androidx.activity.result.contract.ActivityResultContracts;
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
 import androidx.core.util.Consumer;
+import androidx.lifecycle.LifecycleOwner;
 import static androidx.lifecycle.Lifecycle.State.INITIALIZED;
 
+import java.util.Arrays;
+import java.util.List;
 import java.util.Map;
+import java.util.Objects;
 
 /**
  * A launcher for Actions which depend on Permissions.
@@ -27,24 +33,37 @@ import java.util.Map;
  * Therefore instances MUST be created as part of Activity initialization, ideally in the variable init part.
  * See e.g. https://developer.android.com/training/basics/intents/result for more details
  */
-public class PermissionAction {
+public class PermissionAction<T> {
 
-    private final Activity activity;
+    private static final String PARAMETER_KEY = "param_key";
+
+    private final ComponentActivity activity;
     private final PermissionContext permissionContext;
-    private final Consumer<Bundle> callback;
+    private final Consumer<T> callback;
     private final ActivitySavedState savedState;
     private final ActivityResultLauncher<String[]> permissionLauncher;
 
 
     /** creates and registers a permission action with an Activity */
-    public static PermissionAction register(final ComponentActivity activity, final PermissionContext pCtx, final Consumer<Bundle> callback) {
-        if (activity.getLifecycle().getCurrentState() != INITIALIZED) {
-            throw new IllegalStateException("Activity must be in state INITIALIZED when creating launchers: " + activity.getClass().getName());
-        }
-        return new PermissionAction(activity, pCtx, callback);
+    public static <T> PermissionAction<T> register(@NonNull final ComponentActivity activity, @NonNull final PermissionContext pCtx, @Nullable final Consumer<T> callback) {
+        checkActivityObject(activity);
+        return new PermissionAction<>(activity, pCtx, callback);
     }
 
-    private PermissionAction(final ComponentActivity activity, final PermissionContext pCtx, final Consumer<Bundle> callback) {
+    private static void checkActivityObject(@Nullable final ComponentActivity activity) {
+        Objects.requireNonNull(activity);
+        final List<Class<?>> neededClasses = Arrays.asList(LifecycleOwner.class, Context.class);
+        for (Class<?> nc : neededClasses) {
+            if (!nc.isAssignableFrom(activity.getClass())) {
+                throw new IllegalStateException("Subject must implement all of: " + neededClasses + " but doesn't implement: " + nc);
+            }
+        }
+        if (((LifecycleOwner) activity).getLifecycle().getCurrentState() != INITIALIZED) {
+            throw new IllegalStateException("Lifecycle must be in state INITIALIZED when creating launchers: " + activity.getClass().getName());
+        }
+    }
+
+    private PermissionAction(final ComponentActivity activity, final PermissionContext pCtx, final Consumer<T> callback) {
         this.activity = activity;
         this.permissionContext = pCtx;
         this.callback = callback;
@@ -60,7 +79,7 @@ public class PermissionAction {
         launch(null);
     }
 
-    public void launch(final Bundle parameter) {
+    public void launch(final T parameter) {
         launch(parameter, false);
     }
 
@@ -68,7 +87,7 @@ public class PermissionAction {
      * launches the formerly registered permission action with a user-defined parameter.
      * The parameter will be passed to the action code
      */
-    public void launch(final Bundle parameter, final boolean forceSkipPreExplanation) {
+    public void launch(final T parameter, final boolean forceSkipPreExplanation) {
         //check: should we show an explanation BEFORE requesting permission?
         if (!forceSkipPreExplanation && this.permissionContext.shouldShowRequestPermissionRationale(activity)) {
             SimpleDialog.of(activity).setTitle(TextParam.text(permissionContext.getExplanationTitle()))
@@ -79,17 +98,20 @@ public class PermissionAction {
         }
     }
 
-    private void launchPermissionRequest(final Bundle parameter) {
-        this.savedState.set(parameter);
+    private void launchPermissionRequest(final T parameter) {
+        this.savedState.get().putParcelable(PARAMETER_KEY, new ParcelableValue<>().set(parameter));
         this.permissionLauncher.launch(this.permissionContext.getPermissions());
     }
 
     private void handlePermissionResult(final Map<String, Boolean> result) {
         Log.d("PermissionAction result:" + result);
-        final Bundle parameter = this.savedState.get();
+        final ParcelableValue<T> parameterHolder = this.savedState.get().getParcelable(PARAMETER_KEY);
+        final T parameter = parameterHolder == null ? null : parameterHolder.get();
         if (this.permissionContext.hasAllPermissions()) {
             //-> we have permissions now, execute callback
-            callback.accept(parameter);
+            if (callback != null) {
+                callback.accept(parameter);
+            }
             this.savedState.clear();
         } else {
             //-> we still don't have necessary permissions -> show explanation and ask user again

--- a/main/src/main/java/cgeo/geocaching/permission/PermissionContext.java
+++ b/main/src/main/java/cgeo/geocaching/permission/PermissionContext.java
@@ -21,7 +21,9 @@ import java.util.Set;
 public enum PermissionContext {
 
     LOCATION(new String[]{Manifest.permission.ACCESS_COARSE_LOCATION, Manifest.permission.ACCESS_FINE_LOCATION}, R.string.permission_location_explanation, R.string.permission_location_explanation_title),
-    SEARCH_USER_IN_CONTACTS(new String[]{Manifest.permission.READ_CONTACTS}, R.string.permission_contacts_read_explanation, R.string.permission_contacts_read_explanation_title);
+    SEARCH_USER_IN_CONTACTS(new String[]{Manifest.permission.READ_CONTACTS}, R.string.permission_contacts_read_explanation, R.string.permission_contacts_read_explanation_title),
+    SHOW_WALLPAPER(new String[]{Manifest.permission.READ_EXTERNAL_STORAGE}, R.string.permission_read_external_storage_explanation, R.string.permission_read_external_storage_explanation_title);
+
 
 
     private final String[] permissions;

--- a/main/src/main/java/cgeo/geocaching/settings/SettingsActivity.java
+++ b/main/src/main/java/cgeo/geocaching/settings/SettingsActivity.java
@@ -4,6 +4,8 @@ import cgeo.geocaching.Intents;
 import cgeo.geocaching.R;
 import cgeo.geocaching.maps.mapsforge.v6.RenderThemeHelper;
 import cgeo.geocaching.network.AndroidBeam;
+import cgeo.geocaching.permission.PermissionAction;
+import cgeo.geocaching.permission.PermissionContext;
 import cgeo.geocaching.search.BaseSearchSuggestionCursor;
 import cgeo.geocaching.search.BaseSuggestionsAdapter;
 import cgeo.geocaching.search.SearchUtils;
@@ -86,6 +88,8 @@ public class SettingsActivity extends AppCompatActivity implements PreferenceFra
 
     private static final ArrayList<BasePreferenceFragment.PrefSearchDescriptor> searchIndex = new ArrayList<>();
 
+    private final PermissionAction<Void> askShowWallpaperPermissionAction = PermissionAction.register(this, PermissionContext.SHOW_WALLPAPER, null);
+
     @Override
     protected void onCreate(final Bundle savedInstanceState) {
         ApplicationSettings.setLocale(this);
@@ -127,6 +131,10 @@ public class SettingsActivity extends AppCompatActivity implements PreferenceFra
 
     public BackupUtils getBackupUtils() {
         return backupUtils;
+    }
+
+    public void askShowWallpaperPermission() {
+        this.askShowWallpaperPermissionAction.launch();
     }
 
     private void handleIntent(final Bundle savedInstanceState) {

--- a/main/src/main/java/cgeo/geocaching/settings/fragments/PreferenceAppearanceFragment.java
+++ b/main/src/main/java/cgeo/geocaching/settings/fragments/PreferenceAppearanceFragment.java
@@ -4,6 +4,8 @@ import cgeo.geocaching.BuildConfig;
 import cgeo.geocaching.R;
 import cgeo.geocaching.enumerations.QuickLaunchItem;
 import cgeo.geocaching.settings.Settings;
+import cgeo.geocaching.settings.SettingsActivity;
+import static cgeo.geocaching.utils.SettingsUtils.setPrefClick;
 
 import android.os.Bundle;
 
@@ -69,6 +71,13 @@ public class PreferenceAppearanceFragment extends BasePreferenceFragment {
     public void onResume() {
         super.onResume();
         getActivity().setTitle(R.string.settings_title_appearance);
+
+        setPrefClick(this, R.string.pref_wallpaper, () -> {
+            if (Settings.isWallpaper()) {
+                ((SettingsActivity) this.getActivity()).askShowWallpaperPermission();
+            }
+        });
+
     }
 
     private void setLanguageSummary(final ListPreference languagePref, final String newValue) {

--- a/main/src/main/java/cgeo/geocaching/utils/ParcelableValue.java
+++ b/main/src/main/java/cgeo/geocaching/utils/ParcelableValue.java
@@ -1,0 +1,55 @@
+package cgeo.geocaching.utils;
+
+import android.os.Parcel;
+import android.os.Parcelable;
+
+/** A parcelable holder for all values which can be processed by {@link android.os.Parcel#writeValue(java.lang.Object)} */
+public class ParcelableValue<T> implements Parcelable {
+
+    private T value;
+
+    public ParcelableValue() {
+        //empty
+    }
+
+    public ParcelableValue<T> set(final T value) {
+        this.value = value;
+        return this;
+    }
+
+    public T get() {
+        return value;
+    }
+
+    //Parcelable stuff
+
+    @SuppressWarnings("unchecked")
+    protected ParcelableValue(final Parcel in) {
+        this.value = (T) in.readValue(getClass().getClassLoader());
+    }
+
+    @Override
+    public void writeToParcel(final Parcel dest, final int flags) {
+        dest.writeValue(value);
+    }
+
+    @Override
+    public int describeContents() {
+        return 0;
+    }
+
+    @SuppressWarnings("rawtypes")
+    public static final Creator<ParcelableValue> CREATOR = new Creator<ParcelableValue>() {
+
+        @Override
+        public ParcelableValue createFromParcel(final Parcel in) {
+            return new ParcelableValue(in);
+        }
+
+        @Override
+        public ParcelableValue[] newArray(final int size) {
+            return new ParcelableValue[size];
+        }
+    };
+
+}

--- a/main/src/main/res/values/strings.xml
+++ b/main/src/main/res/values/strings.xml
@@ -2442,6 +2442,8 @@
     <string name="permission_contacts_read_explanation">c:geo can search for contacts in your address book using cacher names. To do this, you must authorize c:geo to read your address book. Otherwise, usage of this function is not possible.</string>
     <string name="permission_location_explanation_title">Location permission</string>
     <string name="permission_location_explanation">c:geo is optimized for usage with GPS data (fine location) to locate your position and calculate distance and direction to geocaches. Without this permission, location-based functions will not work as expected.</string>
+    <string name="permission_read_external_storage_explanation_title">Read external Storage permission</string>
+    <string name="permission_read_external_storage_explanation">c:geo can use your systems wallpaper as background for the app. To do this, c:geo needs permission to read your media data on external storage.</string>
 
     <string name="storage_permission_needed">This function is not possible without storage permission.</string>
 

